### ISSUE TITLE
Regarding trainable parameters in equinox

### ISSUE THREAD

**Issue by Basant1861** [patrick-kidger/equinox#710]
I do not have any coding-related issues. My question is regarding the way Equinox treats everything as pytrees. So suppose I have some equinox modules inside my model and let's say I have an additional differentiable optimization layer as well. The parameters/constants of the optimizer are NOT  supposed to be trainable parameters. I wanted to know if Equinox will treat these differentiable optimization layer parameters also trainable. In general, are all constant arrays in the model treated as trainable parameters?I read in the FAQ the following:

> How to mark arrays as non-trainable? (Like PyTorch's buffers?)[¤](https://docs.kidger.site/equinox/faq/#how-to-mark-arrays-as-non-trainable-like-pytorchs-buffers)

>This can be done by using jax.lax.stop_gradient:

>class Model(eqx.Module):
    buffer: Array
    param: Array

  >  def __call__(self, x):
        return self.param * x + jax.lax.stop_gradient(self.buffer)

Does that mean we need to wrap all such constant arrays with jax.lax.stop_gradient?
Regards,

**Comment by lockwo** [comment]
In general, all inexact arrays are differentiable. If there derivative is zero, naturally this is no problem. If their derivative is non zero and you need it to be zero, `stop_gradient` is the way to do

**Comment by patrick-kidger** [comment]
Note also that you can apply `lax.stop_gradient` to any pytree-of-arrays, you don't necessarily need to apply it to each array individually.

**Comment by Basant1861** [comment]
Thanks for the replies guys. A quick question which is actually an extension of the above question. If my differentiable optimization layer is a class which does NOT inherit from equinox.Module(but is part of the model in the sense that it only acts as a transformation of the output from the other layers...but with no intended learnable parameters) then is it correct to say that the differentiable optimization layer would not be treated as a pytree and hence any arrays inside this layer would be treated as non-trainable?

**Comment by patrick-kidger** [comment]
I think that sounds fairly dodgy to be honest! Passing around not-pytrees is an easy way to get JAX tracer leaks.

**Comment by Basant1861** [comment]
I would like to give a slightly concrete example to understand what is happening in my code:

```
class Model(eqx.Module):
    
    x = MLP_1(some_inputs) # inherits from eqx.Module
    x = MLP_2(x)  # inherits from eqx.Module
    x = Diff_Opt_layer(x) ## No learnable parameters in this layer and let's say inherits from eqx.Module
    
```
and finally, compute loss on x.
Then are you saying I should do something like 

`x = jax.lax.stop_gradient(Diff_Opt_layer(x))`
?
Won't this prevent gradient flow through the complete differentiable optimisation layer? Sorry for possibly sounding naive but I do not understand that if I have 100 constant arrays in my optimization layer then how do I apply jax.lax.stop_gradient to prevent these arrays from getting updated during backpropagation? I just do not want these arrays to be treated as learnable parameters.

**Comment by patrick-kidger** [comment]
If I understand your code correctly, you probably want `x = lax.stop_gradient(Diff_Opt_layer)(x)`. That is, remove all of the gradient-tracking for all of the arrays in your `Diff_Opt_layer` pytree.

**Comment by Basant1861** [comment]
Yes, that is exactly what I want! Thanks!

**Comment by patrick-kidger** [comment]
Yup, that's exactly correct.

**Comment by Basant1861** [comment]
Thanks a lot!
